tfjs-backend-webgl/src/webgl.ts

```diff
@@ -35,3 +36,136 @@ export {gpgpu_util, webgl_util};
+ * gl.texParameteri(tex2d, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
+ * gl.texParameteri(tex2d, gl.TEXTURE_MIN_FILTER, gl.NEAREST);
+ * gl.texParameteri(tex2d, gl.TEXTURE_MAG_FILTER, gl.NEAREST);
+ * gl.texImage2D(
+ *   tex2d, 0, gl.R32F, // internalFormat
+ *   width, height, 0,
+ *   gl.RED, // textureFormat
+ *   gl.FLOAT, // textureType
+ *   new Float32Array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]) // data
+ * );
+ *
+ * const logicalShape = [height, width];
+ * const physicalShape = [height, width];
+ * const a = tf.webgl.createTensorFromTexture(texture, logicalShape,
+ *   physicalShape);
+ *
+ * ```
+ *
+ * For postprocessing on the GPU, it's possible to retrieve the texture backing
+ * any tensor by calling the WebGL backend's `getTexture` method like so:
+ *
+ * ```js
+ * const a = tf.tensor1d([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]);
+ * const tex = tf.backend().getTexture(a.dataId);
+ * ```
+ *
+ * @param obj An object with the following properties:
+ *  @param texture The WebGL texture to create a tensor from. The texture must
+ * be unpacked - each texel should only store a single value. The flattened
+ * values of the tensor will be read from left to right, and top to bottom. The
+ * texture can have empty texels at the end, but the values must be written
+ * densely - in other words, all empty texels must be contiguous.
+ *  @param shape The logical shape of the texture.
+ *  @param dtype The dtype of the tensor to be created.
+ *  @param texShapeRC The physical dimensions of the texture expressed as [rows,
+ * columns].
+ *  @param internalFormat The internalFormat of the texture provided to
+ * gl.texImage2D during texture creation.
+ *  @param textureFormat The textureFormat of the texture provided to
+ * gl.texImage2D during texture creation.
+ *  @param textureType The textureType of the texture provided to gl.texImage2D
+ * during texture creation.
+ * @doc {heading: 'Environment', namespace: 'webgl'}
+ */
+export function createTensorFromTexture({
+  texture,
+  shape,
+  dtype,
+  texShapeRC,
+  internalFormat,
+  textureFormat,
+  textureType
+}: TensorFromTextureConfig): Tensor {
+  // OpenGL / WebGL do not make it possible to query textures for their
+  // properties (physical dimensions, internalFormat, etc.), therefore we ask
+  // the user to provide this information in order to validate their texture.
+  // References:
+  // https://stackoverflow.com/questions/30140178/opengl-es-2-0-get-texture-size-and-other-info
+  // https://stackoverflow.com/questions/26315021/is-there-a-way-to-retrieve-the-dimensions-of-a-texture-after-binding-with-gl-bin
+  // https://stackoverflow.com/questions/46387922/how-to-check-a-texture-is-2d-texture-or-cube-texture-in-webgl
+
+  const backend = engine().backend as MathBackendWebGL;
+  const gl = backend.gpgpu.gl;
+  const texConfig = backend.gpgpu.textureConfig;
+  let params: gpgpu_util.TextureCreationParams;
+
+  const isPacked = textureFormat === gl.RGBA;
```

Comment:
Since you require that the texel of input texture must be one value, it's not possible that `textureFormat` is equal to `gl.RGBA`?

In fact, I prefer you support both unpacked texture and the densely rgba texture as the input.

Reply:
Done. Now, we support all non-empty subsequence of 'RGBA'.